x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportFollowStatsAction.java

```diff
@@ -85,7 +98,74 @@ protected void doExecute(
                 throw new ResourceNotFoundException("No shard follow tasks for follower indices [{}]", resources);
             }
         }
-        super.doExecute(task, request, listener);
+        super.doExecute(task, request, listener.map(this::enrichWithIndexDocumentCountLag));
+    }
+
+    private FollowStatsAction.StatsResponses enrichWithIndexDocumentCountLag(final FollowStatsAction.StatsResponses responses)
+        throws Exception {
+        // structure of mapping: follower index -> (remote cluster, leader index)
+        final Map<String, Tuple<String, String>> followerIndexToRemoteCluster = new HashMap<>();
+        responses.getStatsResponses()
+            .stream()
+            .map(FollowStatsAction.StatsResponse::status)
+            .forEach(
+                status -> followerIndexToRemoteCluster.computeIfAbsent(
+                    status.followerIndex(),
+                    ignored -> Tuple.tuple(status.getRemoteCluster(), status.leaderIndex())
+                )
+            );
+        followerIndexToRemoteCluster.forEach((followerIndex, remoteClusterAndIndex) -> {
+            final String remoteCluster = remoteClusterAndIndex.v1();
+            final String leaderIndex = remoteClusterAndIndex.v2();
+            try {
+                responses.setFollowerIndexDocumentCountLag(followerIndex, getDocumentCountLag(followerIndex, leaderIndex, remoteCluster));
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        });
+
+        return responses;
+    }
+
+    private long getDocumentCountLag(final String followerIndex, final String leaderIndex, final String remoteCluster) throws Exception {
```

Comment:
This function will basically make the thread wait for getting the results (at the future's .get()). You use `listener.map()` at the beginning, which basically calls the supplied function synchronously. And inside the function you use these futures, where .get() will wait for them to complete synchronously.

In order to make the whole thing async, you'd ideally need to pass the first `listener` to your functions and complete it after you calculate the new index doc count lag.

Specifically, I imagine enrichWithIndexDocumentCountLag() something like this:

```
ActionListener<FollowStatsAction.StatsResponses> listener enrichWithIndexDocumentCountLag(ActionListener<FollowStatsAction.StatsResponses> finalListener) {

return finalListener.delegateFailure((l, statsResponse) -> {

   // Here we need to enrich the statsResponse before completing the final listener l
   
   try (var listeners = new RefCountingListener(l.map(ignored -> statsResponse))) {
       for (each index) {
           // you get a new listener using listeners.acquire(), and use .runAfter() to get the doc lag integer and .put() it atomically into the statsResponse
           // this new listener, you call getDocumentCountLag() with it as an argument
           // getDocumentCountLag() needs to be restructured to not use futures. You can shoot the two `sendRequests()`, and use a GroupedActionListener (since you already know you just need 2 actions) for them, which once completed you'll calculated the difference of the doc results (be careful to know which one is from the remote and which one from the local transport action), and complete the listener (passed through the function arguments) with the calculated result.
       }
   }

});
```

Ouf, listeners are a difficult art! Please note my idea may not be totally correct, but hopefully you'll get the basic classes/notions that might be useful for your case here.